### PR TITLE
Allow setting BaseComponent Arrays in LoginEvents

### PR DESCRIPTION
**Pull request by Gabscap**
And deprecate `setCancelReason(String)`.

### CODE REVIEW

**Comment by Paulomart** [comment]
This will break plugins using getCancelReason() :: String

**Comment by Gabscap** [comment]
The other option is to seperate these functions by name.
```java
String getCancelReason()
BaseComponent[] getBaseComponentCancelReason()

void setCancelReason(String)
void setBaseComponentCancelReason(BaseComponent[])
```

**Comment by Gabscap** [comment]
Since the client displays the cancelReason in a non-interactable way (in respect to hyperlinks, hover), the LoginEvents stores the cancelReason as a String for backward compatibility, but offers the method `void setCancelReason(BaseComponent[] cancelReason)` to set a BaseComponent array.

File: api/src/main/java/net/md_5/bungee/api/event/LoginEvent.java [hunk 1]
```diff
@@ -34,4 +36,10 @@ public LoginEvent(PendingConnection connection, Callback<LoginEvent> done)
         super( done );
         this.connection = connection;
     }
+
+    @Deprecated
```
Comment (Paulomart) [inline api/src/main/java/net/md_5/bungee/api/event/LoginEvent.java]: Why `Deprecated`

File: api/src/main/java/net/md_5/bungee/api/event/PreLoginEvent.java [hunk 2]
```diff
@@ -39,4 +41,10 @@ public PreLoginEvent(PendingConnection connection, Callback<PreLoginEvent> done)
         super( done );
         this.connection = connection;
     }
+
+    @Deprecated
```
Comment (Paulomart) [inline api/src/main/java/net/md_5/bungee/api/event/PreLoginEvent.java]: Why `Deprecated`

**Comment by Gabscap** [comment]
I changed the commit and forgot to remove it.

**Comment by Gabscap** [comment]
Edited the PR to match the technique used in [ServerKickEvent](https://github.com/SpigotMC/BungeeCord/blob/master/api/src/main/java/net/md_5/bungee/api/event/ServerKickEvent.java).

File: api/src/main/java/net/md_5/bungee/api/event/LoginEvent.java (hunk 1 above)
Comment (Janmm14) [inline api/src/main/java/net/md_5/bungee/api/event/LoginEvent.java]: Because everything and everyone should use BaseComponents

**Comment by md-5** [comment]
I see ServerKick uses arrays, but varargs would be better. I wonder if there is a way to make lombok do that.

**Comment by Mystiflow** [comment]
Also linked to #1336

**Comment by Gabscap** [comment]
The Setter now uses varags.

File: api/src/main/java/net/md_5/bungee/api/event/PreLoginEvent.java [hunk 3]
```diff
@@ -39,4 +44,20 @@ public PreLoginEvent(PendingConnection connection, Callback<PreLoginEvent> done)
...
+        return BaseComponent.toLegacyText( cancelReasonComponent );
+    }
+
+    @Deprecated
+    public void setCancelReason(String reason)
+    {
+        cancelReasonComponent = TextComponent.fromLegacyText( reason );
+    }
+
+    public void setCancelReason(BaseComponent... reason) {
```
Comment (Mystiflow) [inline api/src/main/java/net/md_5/bungee/api/event/PreLoginEvent.java]: setCancelReasonComponent to match kickReasonComponent?

Comment (Gabscap) [inline api/src/main/java/net/md_5/bungee/api/event/PreLoginEvent.java]: Right. Just a second

**Comment by md-5** [comment]
Was a tough compromise, but ultimately went with  mostly this in ea66802.

**Pull request closed without merging**